src/tools/wasm-merge.cpp

```diff
@@ -0,0 +1,527 @@
+  std::vector<std::string> inputFiles;
+  std::vector<std::string> inputFileNames;
+  bool emitBinary = true;
+  bool debugInfo = false;
+
+  const std::string WasmMergeOption = "wasm-merge options";
+
+  ToolOptions options("wasm-merge",
+                      R"(Merge wasm files into one.
+
+For example,
+
+  wasm-merge foo.wasm foo bar.wasm bar -o merged.wasm
+
+will read foo.wasm and bar.wasm, with names 'foo' and 'bar' respectively, so if the second imports from 'foo', we will see that as an import from the first module after the merge. The merged output will be written to merged.wasm.
+
+Note that filenames and modules names are interleaved (which is hopefully less confusing).)");
+
+  options
+    .add("--output",
+         "-o",
+         "Output file (stdout if not specified)",
+         WasmMergeOption,
+         Options::Arguments::One,
+         [](Options* o, const std::string& argument) {
+           o->extra["output"] = argument;
+           Colors::setEnabled(false);
+         })
+    .add_positional("INFILE1 NAME1 INFILE2 NAME2 [..]",
+                    Options::Arguments::N,
+                    [&](Options* o, const std::string& argument) {
+                      if (inputFiles.size() == inputFileNames.size()) {
+                        inputFiles.push_back(argument);
+                      } else {
+                        inputFileNames.push_back(argument);
+                      }
+                    })
+    .add("--emit-text",
+         "-S",
+         "Emit text instead of binary for the output file",
+         WasmMergeOption,
+         Options::Arguments::Zero,
+         [&](Options* o, const std::string& argument) { emitBinary = false; })
+    .add("--debuginfo",
+         "-g",
+         "Emit names section and debug info",
+         WasmMergeOption,
+         Options::Arguments::Zero,
+         [&](Options* o, const std::string& arguments) { debugInfo = true; });
+  options.parse(argc, argv);
+
+  if (inputFiles.size() != inputFileNames.size()) {
+    Fatal() << "Please provide an import name for each input file. "
+               "In particular, the number of positional inputs must be even as "
+               "each wasm binary must be followed by its name.";
+  }
+
+  // Process the inputs.
+  // TODO: If the inputs are a very large number of small modules then it might
+  //       make sense to parallelize this. (If so, then changing the existing
+  //       parallelism above in NameMapper might make sense.)
+
+  bool first = true;
+  for (Index i = 0; i < inputFiles.size(); i++) {
+    auto inputFile = inputFiles[i];
+    auto inputFileName = inputFileNames[i];
+
+    if (options.debug) {
+      std::cerr << "reading input '" << inputFile << "' as '" << inputFileName
+                << "'...\n";
+    }
+
+    // For the first input, we'll just read it in directly. For later inputs,
+    // we read them and then merge.
+    std::unique_ptr<Module> laterInput;
+    Module* currModule;
+    if (first) {
```

Comment:
It might be simpler to just use `i == 0` instead of a separate `first` variable.